### PR TITLE
Set default CMAKE_BUILD_TYPE to Release and implement CMAKE_BUILD_TYPE FastBuild

### PR DESCRIPTION
**Pull request by marbre**
* Sets the default `CMAKE_BUILD_TYPE` to `Release`
* Implements a custom build type `FastBuild`
  * If the `CMAKE_BUILD_TYPE` is set to `FastBuild`, for LLVM the `CMAKE_BUILD_TYPE` is set to `Relase`

### CODE REVIEW

**Comment by GMNGeoffrey** [comment]
So it looks like this increases the build time by another third (40 minutes, vs 30 minutes for the last run on master: https://github.com/google/iree/commit/fea7914fdd77e93976c055c366cd9d8373fae1d0/checks?check_suite_id=399128460). That surprises me. I would have expected it to be faster than debug. Should we reconsider making release the default? Maybe we should default to fastbuild

**Comment by marbre** [comment]
Well regarding the default in the CMakeLists.txt, I think Release is the option to go. Anyway, fell free to pass `-DCMAKE_BUILD_TYPE=FastBuild` within the build pipeline. At least on my build machine #475 seems to be no issue with CMake.

**Comment by GMNGeoffrey** [comment]
Why do you say Release is better? You're probably the main one building with cmake, so I'm fine deferring to your judgement on that, but curious about your reasoning. 

Just doing a rough and unscientific test using `time` on my workstation using the same script as the CI

```shell
cmake \
  -DCMAKE_BUILD_TYPE=${BUILD_TYPE} \
  -DIREE_BUILD_COMPILER=ON \
  -DIREE_BUILD_TESTS=ON \
  -DIREE_BUILD_SAMPLES=OFF \
  -DIREE_BUILD_DEBUGGER=OFF \
  ..
```

With BUILD_TYPE=
Release: 19m58s
FastBuild: 17m19s
Debug: 15m22s

So for strange reasons, Debug is actually even faster than FastBuild and it seems like things are about twice as fast on my workstation as on the GitHub runner. I guess if we start actually running tests, this balance would change quite a bit, but it seems like while we're just trying to get as much building as possible that the fastest build is the most desirable? It also calls into question somewhat the utility of fastbuild at all.

**Comment by marbre** [comment]
Well, normally I build software with optimizations enabled, since I am not scared about compile time but rather about execution time. So I assume that Release would be the suitable default for a user.

Regarding compile time, setting `FastBuild`, LLVM is actually build with `CMAKE_BUILD_TYPE=Release`, so with optimizations. With `Debug` set, IREE as well as LLVM are build without optimizations. I think that is the reason for `FastBuild` beeing build slower than `Debug`.

**Comment by GMNGeoffrey** [comment]
Fair enough. For now, since we're not actually running anything based on the build, let's at least default the CI to something that builds quickly. Can you change that as part of this PR? (build_tools/scripts/cmake_build.sh)

Maybe for FastBuild we should build LLVM with Debug if that's faster? I'm testing that out now to see how it looks.

**Comment by marbre** [comment]
Sure, I can adjust cmake_build.sh :) However, the question remains if we only want to keep 694d681 and set the CI to `Debug` in addition, but skip the remaining commits for `FastBuild`. If it just blows up the CMakeLists.txt without any benefit I think we should rather drop those changes.

**Comment by GMNGeoffrey** [comment]
Well I forgot to actually run it with `time` :man_facepalming:, so all I've got is the timestamps I have printed as part of my prompt. Running again, but it does look like FastBuild with LLVM set to Debug is slightly faster than everything Debug (14m). I'm fine doing that for now. Or if you'd prefer, we can just drop FastBuild and stick with default Release but CI using Debug

**Comment by GMNGeoffrey** [comment]
Yeah running with time, FastBuild setting LLVM to Debug is the fastest by a about 30 seconds (14m50s). Not a huge difference though.

**Comment by marbre** [comment]
Okay, done. The CI should build with `Debug`, which is LLVM's default for now as well. Thus I think it might be okay to keep the `FastBuild` setting for now. We can drop it later anyway if there isn't much benefit, but at least it makes the CMake and Bazel builds more comparable.

File: build_tools/scripts/cmake_build.sh [hunk 1]
```diff
@@ -26,6 +26,6 @@ CMAKE_BIN=${CMAKE_BIN:-$(which cmake)}
 cd ${ROOT_DIR?}
 rm -rf build/
 mkdir build && cd build
-"$CMAKE_BIN" -DIREE_BUILD_COMPILER=ON -DIREE_BUILD_TESTS=ON -DIREE_BUILD_SAMPLES=OFF -DIREE_BUILD_DEBUGGER=OFF ..
+"$CMAKE_BIN" -DCMAKE_BUILD_TYPE=Debug -DIREE_BUILD_COMPILER=ON -DIREE_BUILD_TESTS=ON -DIREE_BUILD_SAMPLES=OFF -DIREE_BUILD_DEBUGGER=OFF ..
```
Comment (GMNGeoffrey) [inline build_tools/scripts/cmake_build.sh]: I was saying I think this should be FastBuild

```suggestion
"$CMAKE_BIN" -DCMAKE_BUILD_TYPE=FastBuild -DIREE_BUILD_COMPILER=ON -DIREE_BUILD_TESTS=ON -DIREE_BUILD_SAMPLES=OFF -DIREE_BUILD_DEBUGGER=OFF ..
```

Comment (marbre) [reply build_tools/scripts/cmake_build.sh]: You're absolutely right. Sorry for that, as an excuse, it's quite late here.

Comment (marbre) [reply build_tools/scripts/cmake_build.sh]: Oh, I could have picked that directly here on GitHub. Wasn't aware of this feature. Next time :)

**Comment by GMNGeoffrey** [comment]
Build times on the CI look better now, thanks!

**Pull request closed without merging**